### ISSUE TITLE
Bug: Data Collection Fails with TF401174

### ISSUE THREAD

**Issue by paulmccrady** [cloudyspells/PSRule.Rules.AzureDevOps#53]
# **🐞 Bug Report**

## **Describe the bug**
Running Export-AzDevOpsRuleData on a new repo, with a new pipeline not on the main/default branch fails with the following error:

Invoke-RestMethod: /Users/paul.mccrady@contino.io/PSRule.Rules.AzureDevOps/src/PSRule.Rules.AzureDevOps/Functions/DevOps.Pipelines.Core.ps1:223:21
Line |
 223 |  …  $response = Invoke-RestMethod -Uri $uri -Method Get -Headers $header
     |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | {"$id":"1","innerException":null,"message":"TF401174: The item 'test-pipeline.yaml' could not be found in the
     | repository 'PmcTest' at the version specified by '' (resolved to commit
     | '97cc80fe33624aeff34d81adb6a4a409aabf69bf')","typeName":"Microsoft.TeamFoundation.Git.Server.GitItemNotFoundException,
     | Microsoft.TeamFoundation.Git.Server","typeKey":"GitItemNotFoundException","errorCode":0,"eventId":3000}

---

### **Is this a regression?**
I don't know.

---

### **To Reproduce**

1. Create a new repo in azure devops with just a readme file
2. Create a new branch of master/main
3. Add a test yaml pipeline and commit (note - DO NOT merge to master/main)
4. Create a pipeline in azure devops referencing the yaml above on the new branch
5. Run Export-AzDevOpsRuleData

---

### **Expected behaviour**

Data Collection should run without throwing an error.

---

### **Media prove**

![Screenshot 2023-10-26 at 09 26 39](https://github.com/cloudyspells/PSRule.Rules.AzureDevOps/assets/30175621/0772fe72-747c-4f25-9074-ac1646bdfb78)

![Screenshot 2023-10-26 at 09 27 30](https://github.com/cloudyspells/PSRule.Rules.AzureDevOps/assets/30175621/8aeb709b-d43f-489c-a199-1e88e72b9375)

![Screenshot 2023-10-26 at 09 28 16](https://github.com/cloudyspells/PSRule.Rules.AzureDevOps/assets/30175621/d97b7d4b-9e06-4058-8591-17e54a18a203)

![Screenshot 2023-10-26 at 09 29 08](https://github.com/cloudyspells/PSRule.Rules.AzureDevOps/assets/30175621/5f6e961f-02dc-4949-afd1-837ae1fbe547)

---

### **Your environment**

OS: Darwin 22.6.0 Darwin Kernel Version 22.6.0
PowerShell version: 7.3.3
PSRule version: 2.9.0
PSRule.Rules.AzureDevOps version: v0.2.0

---

### **Additional context**

Export command was run from my dev mac using a ReadOnly PAT.

**Comment by webtonize** [comment]
Hi @paulmccrady, thanks so much for another well described report. This really helps improve the module.

I checked the output the module currently retrieves from the Azure DevOps REST API and that only contains a reference to the YAML definition file name/path but I do not see any git branch reference I can use in the retrieval of the YAML file. I am still exploring other API endpoint to query to get information on a branch reference for the pipeline definition, but have not yet been able to find one. So we might have to consider pipelines not yet defined in the repositories' default branch as non-auditable. Contributions are very welcome here.

**Comment by paulmccrady** [comment]
Get a list of branches using 
https://dev.azure.com/{organisation}/{project}/_apis/git/repositories/{repositoryId}/refs?api-version=7.2-preview.1

Loop through each returned branch (starting with the default branch), use the api to search for the item (like you have done) but supply the versionDescriptor.version param (below shows a branch called test ...

https://dev.azure.com/{organisation}/{project}/_apis/git/repositories/{repositoryId}/items?path=test-pipeline.yaml&versionDescriptor.version=test&api-version=7.2-preview.1

This returned the yaml

`trigger:
- none

pool:
    vmImage: 'ubuntu-latest'

steps:
- task: Bash@3
  displayName: 'Test echo'
  inputs:
    targetType: inline
    script: |
      echo "Hello from a simple task"
`

Once I'm up and running I'll help out. Above solution isn't great I admit, but it's not an each one to solve.

**Comment by webtonize** [comment]
As a fix for the hard error, I have adjusted the module to warn the user the pipeline definition was not found in the default branch and thus could not be exported. 

Although I do think it would be better to expand the modules view a bit beyond just the default branch, I also think that we should take some broader structured approach in doing so. By just looping through the branches we really don't know how relevant the version is in relation to the other checks, like branch policies etc. What I would like to do is that the user can configure some patterns for branch names that should be regarded as up to standard. We could then also use those branch name patterns within the rules to check that no code in other branches has ways of reaching production (code) without the proper controls in between. That would require some significant development effort, so therefore I chose to stick with the default branch solution for the time being in this issue.
